File: DemoClient.cs
using Demo.V1;
using Google.Protobuf.WellKnownTypes;
using Grpc.Net.Client;

var serverAddr = Environment.GetEnvironmentVariable("SERVER_ADDR") ?? "http://server:50051";

// Ensure scheme is present for GrpcChannel
if (!serverAddr.StartsWith("http"))
{
    serverAddr = "http://" + serverAddr;
}

using var channel = GrpcChannel.ForAddress(serverAddr);
var client = new DemoService.DemoServiceClient(channel);

// Echo RPC
var echoResponse = await client.EchoAsync(new EchoRequest { Message = "hello from csharp client" });
Console.WriteLine($"Echo response: {echoResponse.Message}");

// Health RPC
var healthResponse = await client.HealthAsync(new Empty());
Console.WriteLine($"Health response: {healthResponse.Status}");

File: demo_client.ts
import * as grpc from '@grpc/grpc-js';
import { GrpcTransport } from '@protobuf-ts/grpc-transport';
import { DemoServiceClient } from './gen/demo/v1.client.js';
import { EchoRequest } from './gen/demo/v1.js';
import { Empty } from './gen/google/protobuf/empty.js';

const addr = process.env.SERVER_ADDR ?? 'server:50051';
const transport = new GrpcTransport({
  host: addr,
  channelCredentials: grpc.credentials.createInsecure(),
});
const client = new DemoServiceClient(transport);

const { response: echoRes } = await client.echo(EchoRequest.create({ message: 'Hello from TypeScript!' }));
console.log(`Echo response: ${echoRes.message}`);

const { response: healthRes } = await client.health(Empty.create());
console.log(`Health response: ${healthRes.status}`);

transport.close();
